### DISCUSSION TITLE
Using the same resource on several tasks at once

### DISCUSSION THREAD

**Discussion by marc-hg** [pybricks/support#1141] [Q&A] [answered]
Recently I've been playing around with multitasking and came across the following error:

> _OSError: This resource cannot be used in two tasks at once._

When trying to use _hub.speaker.play_notes_ on two different methods running at once through _multi_task_.

I understand this is an intended exception but I was wondering if there's perhaphs a flag under multi_task method where if two code executions try to access the same resource at once it just blocks the second or overwrites the first one. 

Or, if alternatively, this is something I should do myself with some wrapper logic over the resources (for example a wrapper over play_notes)

Here's a sample program to show this exception. It triggers when pressing the two hub buttons at the same time:

```python
from pybricks.hubs import PrimeHub
from pybricks.tools import wait, multitask, run_task
from pybricks.parameters import Button

hub = PrimeHub()
tasks = []
events = []

def press_right_condition():
  return Button.RIGHT in hub.buttons.pressed()
async def press_right_action():
  await hub.speaker.play_notes(['C4/4'], 240)
events.append((press_right_condition, press_right_action))

def press_left_condition():
  return Button.LEFT in hub.buttons.pressed()
async def press_left_action():
  await hub.speaker.play_notes(['D4/4'], 240)
events.append((press_left_condition, press_left_action))

async def event_task(condition, action):
    ready_to_trigger = True
    while True:
        is_condition_met = condition()
        if is_condition_met and ready_to_trigger:
            await action()
        ready_to_trigger = not is_condition_met
        await wait(10)

tasks += [event_task(event, trigger) for event, trigger in events]
async def main():
    await multitask(*tasks)

# Run the main program
run_task(main())
```

**Comment by dlech** [comment, marked as answer]
It sounds like you want a lock similar to `asyncio.Lock` from CPython. We don't have a builtin one, but it is easy enough to add your own implementation. Here is a trivial implementation:

```python
class AsyncLock:
    def __init__(self):
        self._held = False

    def __aenter__(self):
        while self._held:
            yield

        self._held = True

        return self

    async def __aexit__(self, t, e, s):
        self._held = False
```

More complex programs might need a "fair" lock that maintains a queue of waiting tasks so that they are processed in the same order that they requested the lock, but for a program like this, the naive implementation is probably good enough.

Here is how to use it (I took some liberties with your program - nothing wrong with the way you did it - I just wanted to try some other things out).

```python
from pybricks.hubs import PrimeHub
from pybricks.tools import wait, multitask, run_task
from pybricks.parameters import Button

EVENT_MAP = {
    Button.RIGHT: 'C4/4',
    Button.LEFT: 'D4/4',
}

hub = PrimeHub()

class AsyncLock:
    def __init__(self):
        self._held = False

    def __aenter__(self):
        while self._held:
            yield

        self._held = True

        return self

    async def __aexit__(self, t, e, s):
        self._held = False


speaker_lock = AsyncLock()

def press_condition(button):
    def test():
        return button in hub.buttons.pressed()
    
    return test

def press_action(note):
    async def play_note():
        async with speaker_lock:
            await hub.speaker.play_notes([note], 240)
    
    return play_note

async def event_task(condition, action):
    ready_to_trigger = True

    while True:
        is_condition_met = condition()

        if is_condition_met and ready_to_trigger:
            await action()

        ready_to_trigger = not is_condition_met

        await wait(10)

async def main():
    await multitask(
        *(event_task(press_condition(button), press_action(note)) 
        for button, note in EVENT_MAP.items())
    )

# Run the main program
run_task(main())
```

**Reply by marc-hg** [reply]
Thanks a lot for the clear example. This answers my question.

Do you know by any chance if it's also possible to overwrite the resource rather than waiting for it to be unlocked? For example, if play_note is being used by a task, and another task calls it simultaneously, instead of waiting for it to finish to make the first sound stop and play the second one

**Reply by laurensvalk** [reply]
That is how it currently works for motors: the latest command wins and the other awaitables are cancelled.

Maybe we should consider it for sounds too?

**Reply by marc-hg** [reply]
I see, thanks. It's good to know that's the behaviour on motors.

I understand then that there's currently no way of stopping a sound instruction once it's running?

**Reply by dlech** [reply]
> Do you know by any chance if it's also possible to overwrite the resource rather than waiting for it to be unlocked?

In theory, I think this should work. However, we seem to have a bug where `await speaker_task` still waits for a long time after we `close()` the task (probably waiting for the next task?).

```python
from pybricks.hubs import PrimeHub
from pybricks.tools import wait, multitask, run_task
from pybricks.parameters import Button

EVENT_MAP = {
    Button.RIGHT: 'C4/4',
    Button.LEFT: 'D4/4',
}

hub = PrimeHub()

class TaskInterlock:
    def __init__(self):
        self._task = None

    def set(self, task):
        self._task = task

    def reset(self):
        if self._task is not None:
            self._task.close()

        self._task = None


speaker_interlock = TaskInterlock()

def press_condition(button):
    def test():
        return button in hub.buttons.pressed()
    
    return test

def press_action(note):
    async def play_note():
        speaker_interlock.reset()
        speaker_task = hub.speaker.play_notes([note], 40)
        speaker_interlock.set(speaker_task)
        await speaker_task
    
    return play_note

async def event_task(condition, action):
    ready_to_trigger = True

    while True:
        is_condition_met = condition()

        if is_condition_met and ready_to_trigger:
            await action()

        ready_to_trigger = not is_condition_met

        await wait(10)

async def main():
    await multitask(
        *(event_task(press_condition(button), press_action(note)) 
        for button, note in EVENT_MAP.items())
    )

# Run the main program
run_task(main())
```

**Reply by laurensvalk** [reply]
> Do you know by any chance if it's also possible to overwrite the resource rather than waiting for it to be unlocked? For example, if play_note is being used by a task, and another task calls it simultaneously, instead of waiting for it to finish to make the first sound stop and play the second one

Since this was a very new function that was not released yet, I think we can change the configuration that sounds can be cancelled (by starting a new sound) just like motors. This is done now and should be available in the next beta release.

**Reply by marc-hg** [reply]
Thanks to both!

I think it makes more sense for the behaviour to be like that by default. Not just for consistency with motors but it feels more intuitive and wouldn't break the program if any user has some complex program that plays sounds on different multi tasks. As another positive point one could make use of it to stop currently playing sounds by playing a sound at 0 volume for 0.1s for example

**Reply by dlech** [reply]
FYI, we just released a new beta today with the fix if you want to try it out. You should be able to use `speaker.beep(0, 0)` to stop other sounds.

**Reply by marc-hg** [reply]
Been trying it today and works like a charm. Thanks for notifying
